### PR TITLE
docs: add MQTT Proposal Draft v0.1.0

### PR DESCRIPTION
**Pull request by shantanoo-desai**
- Add detailed description of possible Integration
  of Akri and MQTT
- Add diagrams for different descriptions

**What this PR does / why we need it**:

This PR introduces an initial draft to integrate MQTT Protocol to Akri. 
Since there is already a discussion thread for Support of MQTT (see #106 ), this proposal builds upon the discussion and helps
provide a more detailed description of the protocol under question (MQTT) and how a possible integration might look like.

This documentation is written while considering some already provided proposals under the `docs/proposals` directory.
It contains some questions pertaining to implementation as well as integration aspects of MQTT with Akri.

This proposal is bound to be edited and changed based on discussions/feedback/criticisms.

**Special notes for your reviewer**:

This PR should not break any existing Test Pipelines. 

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] added code adheres to standard Rust formatting (`cargo fmt`)
- [ ] code builds properly (`cargo build`)
- [ ] code is free of common mistakes (`cargo clippy`)
- [ ] all Akri tests succeed (`cargo test`)
- [ ] inline documentation builds (`cargo doc`)
- [ ] version has been updated appropriately (`./version.sh`)

### CODE REVIEW

File: docs/proposals/mqtt.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+Agent implements [MQTT Protocol](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html), a light-weight
+publish/subscribe messaging protocol widely used in Internet of Things (IoT).
+
+Supporting MQTT as a protocol in Akri provides a mechanism by which Application Developers (IoT, Kubernetes)
+can provide scalable applications based on current status of the IoT Devices within an environment. 
+
+Depending on the IoT Device's availability the application pertaining to the data such a device produces
+can be scaled accordingly.
+
+## What can Akri solve?
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:13]: Should this be narrowed based on the discovery strategy. For example, Akri + MQTT Broker discovery makes it so an operator can say, run this client (Akri "broker") on nodes that can see an MQTT Broker with x,y, and z topics. How do MQTT clients normally respond to topics no longer being published (devices going away)? It seems like with this implementation, if a topic goes away that a client/akri broker was using, Akri would automatically bring down the client. Is that ideal?

File: docs/proposals/mqtt.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+> NOTE: MQTT is completely data-agnostic. The Payload and Topic strings can be developed according to the application developer's
+        perspective.
+
+
+## Discovery Process
+
+Since MQTT bases itself over the TCP/IP stack for networking, the specifications provide _handshaking_ mechanisms which can be leveraged
+to discover IoT Devices that communicate with the MQTT Broker. Since the only mediating entity for message exchange is the MQTT Broker,
+all client only require the IP Address and the MQTT port as a basic form of initiation of communication. Default port for unsecure MQTT
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:132]: nit: all clients
Also, maybe clarify that the clients require the "IP address and MQTT port **of the broker** as a ..."

File: docs/proposals/mqtt.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+Broker is 1883 and for a secure Broker is 8883.
+
+The Akri Discovery Handler hence needs to have the URL Information for the MQTT Brokers and particular topics that need to be subscribed to
+in a configuration file.
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:136]: This is an Akri Configuration right? or some other configuration?

File: docs/proposals/mqtt.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+
+## Akri's Custom Broker Interfacing with an MQTT Broker Server
+
+> Note: due to conflicting naming for Akri's Broker as well as MQTT Broker, we refer to Akri's custom broker for MQTT as **MQTT Monitoring Broker**
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:141]: Great note here!

File: docs/proposals/mqtt.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+5. If the IoT Devices already have been publishing data prior to connection of Monitoring Broker with an assumption that they publish their data with QoS 1,
+   the Broker Server would publish all the information to the Monitoring Broker (persistent storage). If IoT Devices are yet to publish on their dedicated
+   topics, the MQTT Broker Server will publish data to the Monitoring Broker once the data is available.
+
+A possible integration of Akri's MQTT Monitoring Broker and MQTT Broker Server looks as follows:
+
+![Akri MQTT Monitoring Broker and MQTT Broker Server System Diagram](../media/akri_mqtt_broker_server.png)
+
+
+## Questions
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:188]: To add to the questions, for each device Akri discovers, a kubernetes resource and an Akri Instance are made. For example, for udev with a filtering rule to only find usb cameras, a resource and Instance would be created for each instance. Then Pods can request these resources. Akri automates the deployment of Pods ("brokers") requesting the resources. What would be a kubernetes resource? (1) each discovered MQTT broker (2) each topic of each MQTT broker (3) MQTT broker with a set of topics (4) something else? How do we make it so a Pod can request the resource(s) that represent "a MQTT broker with a thermometer topic"?

File: docs/proposals/mqtt.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+- Would a generic MQTT Monitoring Broker be possible to implement with Akri?
+    - If not, maybe would it be wise to begin with a Specification like [Sparkplug MQTT Specification](https://sparkplug.eclipse.org) where the Topic Structure
+      and the Payload are defined?
+
+- Is gRPC Server required within the MQTT Monitoring Broker? Does the Publish/Subscribe mechanism require a different technology?
+
+- How actually would Discovery Handler Look like? When one talks about Discovery Handler, is this pertaining to only MQTT Broker Server or the discovery of leaf
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:196]: It seems like this should have its own section in the proposal with the following question also seeming to fit in a "What does discovery mean?" section. From previous discussions, it seems like Akri's role is discovering MQTT Broker Servers.

File: docs/proposals/mqtt.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+```yaml
+apiVersion: akri.sh/v0
+kind: Configuration
+metadata:
+  name: mqtt
+spec:
+  discoveryDetails:
+    name: mqtt
+    discoveryDetails: |+
+      serverUrls:
```
Comment (kate-goldenring) [inline docs/proposals/mqtt.md:220]: What are the server urls? The urls to the MQTT brokers? Is there a way to discover them without being given their addresses? If you are modeling after OPC UA, the server urls can be explicitly provided but ideally that implementation will be expanded to do a network scan instead. Is something like that possible for MQTT?

**Review by kate-goldenring — COMMENTED** [review]
Thank you for this awesome work. The background and diagrams are really helpful. My main questions that i think could be fleshed out in the proposal are:
What resources should Akri advertise?
Why is this an important scenario? As you pose in the question section, is this only useful on the Edge?
What does discovery look like? Network scan for MQTT brokers?
What language do you think you would implement the discovery handler in?

**Comment by bfjelds** [comment]
It would help me to see some of the Akri concepts identified explicitly.  For example, how does the MQTT Monitoring Broker or MQTT Custom Application relate to Akri's Discovery Handler or Broker concepts (is it one or the other or both)?  

I'd also like a section to identify what existing challenge an Akri-MQTT solution would make easier.  For example, I could imagine a couple:

* Use the Akri-MQTT-Discovery-Handler to determine which nodes to deploy an MQTT kubernetes application to ... this approach would make the MQTT kuberenetes application an Akri broker (and the Akri configuration would likely not create any Akri services)
* Bring MQTT data to Kubernetes applications (applications that do not have MQTT client code in them) ... this approach would make the Akri broker truly a broker, converting the MQTT data stream into something that is consumed through the Akri services.

File: docs/proposals/mqtt.md (hunk 7 above)
Comment (shantanoo-desai) [reply docs/proposals/mqtt.md:220]: There is no implementation MQTT Broker discovery. The only possible way to achieve this is AFAIK is through [Zeroconf/mDNS/DNS-SD Solution](https://github.com/mqtt/mqtt.org/wiki/broker_auto-discovery). Possible scan on standard ports like 1883 or 8883 might yield results of possible MQTT Brokers on devices however this is not standard practice.

There is however an alternative Protocol called [MQTT for Sensor Networks aka. MQTT-SN](https://www.oasis-open.org/committees/download.php/66091/MQTT-SN_spec_v1.2.pdf) which is an OASIS Standard which tackles the issue of making MQTT-SN more compatible for Low Power Devices in Lossy Networks that do not necessarily use TCP/IP Stack e.g. ZigBee, Bluetooth, LoRa etc.

File: docs/proposals/mqtt.md (hunk 1 above)
Comment (shantanoo-desai) [reply docs/proposals/mqtt.md:13]: > How do MQTT clients normally respond to topics no longer being published (devices going away)?
There is a possibility of maintaining the connection session between the akri broker (containing mqtt client) however to my knowledge there is no mechanism in MQTT that provides a graceful disconnection if there is no data being published on the subscribed topics. I have to the check the MQTTv5 Specifications where they provide [Session Expiry and Message Expiry](https://www.hivemq.com/blog/mqtt5-essentials-part4-session-and-message-expiry/).
The Session Expiry feature might be able to trigger a graceful disconnection, which in turn leads to bringing down the Akri broker in a much more dynamic manner.

**Comment by shantanoo-desai** [comment]
@kate-goldenring @bfjelds Closing this PR at the moment since I am stuck with some work and will come back to it once I dedicate some time for Akri. If there are some members willing to take this topic, I would gladly assist

**Comment by kate-goldenring** [comment]
> @kate-goldenring @bfjelds Closing this PR at the moment since I am stuck with some work and will come back to it once I dedicate some time for Akri. If there are some members willing to take this topic, I would gladly assist

@shantanoo-desai sounds good. Thanks for the update!

**Pull request closed without merging**